README.md
# Storybook Addon Material-UI Mode

An addon that provide UI to toggle state between light/dark.

![image](https://user-images.githubusercontent.com/18292247/117999815-6f996e00-b36f-11eb-9504-659bbe249e3c.png)

> Works with material-ui v4, v5

## Usage

Create decorator in `.storybook/preview.js` or inside a story.

```js
// .storybook/preview.js
import { useMemo } from "react";
import { MuiThemeProvider, createMuiTheme } from "@material-ui/core/styles";
import CssBaseline from "@material-ui/core/CssBaseline";

const withMuiTheme = (Story, context) => {
  const mode = context.globals?.muiMode;
  const theme = useMemo(() => createMuiTheme({
    palette: {
      mode, // for v5
      // type: mode // for v4
    }
  }), [mode])
  return (
    <MuiThemeProvider theme={theme}>
      <CssBaseline />
    </MuiThemeProvider>
  )
}
```

CHANGELOG.md
# v0.0.14 (Sun Sep 19 2021)

#### ⚠️ Pushed to `main`

- fix: prevent setData after unmount ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.13 (Sun Sep 19 2021)

#### ⚠️ Pushed to `main`

- feat: add googleFontSecondary and svg styling ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.12 (Sun Sep 19 2021)

#### ⚠️ Pushed to `main`

- feat: update mui logo ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.11 (Sat Jun 26 2021)

#### ⚠️ Pushed to `main`

- fix: use weight 300 as fallback ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.10 (Sat Jun 26 2021)

#### ⚠️ Pushed to `main`

- fix: http to https font url ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.9 (Sat Jun 26 2021)

#### ⚠️ Pushed to `main`

- fix: remove default font value ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.8 (Fri Jun 25 2021)

#### ⚠️ Pushed to `main`

- fix: check head exists ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.7 (Fri Jun 25 2021)

#### ⚠️ Pushed to `main`

- fix: make fn async ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.6 (Fri Jun 25 2021)

#### ⚠️ Pushed to `main`

- fix: move apikey to env and fetch all font weight ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.5 (Fri Jun 25 2021)

#### ⚠️ Pushed to `main`

- fix: also add font to iframe head ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.4 (Fri Jun 25 2021)

#### ⚠️ Pushed to `main`

- feat: add GoogleFontSelect ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.3 (Thu Jun 24 2021)

#### ⚠️ Pushed to `main`

- feat: add GoogleFontSelect ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.2 (Wed May 12 2021)

#### ⚠️ Pushed to `main`

- fix: remove log ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))

---

# v0.0.1 (Wed May 12 2021)

#### ⚠️ Pushed to `main`

- feat: ready for release ([@siriwatknp](https://github.com/siriwatknp))
- Initial commit ([@siriwatknp](https://github.com/siriwatknp))

#### Authors: 1

- Siriwat K ([@siriwatknp](https://github.com/siriwatknp))


src/components/GoogleFontSelect.js
import React from "react";
import Select, { components } from "react-select";

const createStyleElement = (font) => {
  const weight = font.files.regular || font.files[300];
  const styleElm = document.createElement("style");
  if (!weight) {
    return styleElm;
  }
  styleElm.innerHTML = `
    @font-face {
      font-family: '${font.family}';
      font-style: normal;
      font-weight: 400;
      font-display: block;
      src: url(${weight.replace("http:", "https:")}) format('truetype');
    }
  `;
  return styleElm;
};

const appendRegularFontToHead = (font) => {
  // add to root head
  document
    .getElementsByTagName("head")[0]
    .appendChild(createStyleElement(font));

  // add to preview-iframe
  const iframe = document.getElementById("storybook-preview-iframe");
  if (iframe) {
    iframe.contentDocument.head.appendChild(createStyleElement(font));
  }
};

function debounce(func, timeout = 100) {
  let timer;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      func.apply(this, args);
    }, timeout);
  };
}

const OPTION_HEIGHT = 24;
const LIST_HEIGHT = 300;
const RANGE = Math.ceil(300 / 24);

const MenuList = ({ innerRef, innerProps, ...props }) => {
  const menuListRef = React.useRef();
  const cacheRef = React.useRef({});

  const injectStyle = (menuListDOM) => {
    const scrollTop = menuListDOM.scrollTop;
    const middleItemIndex = Math.ceil(
      (scrollTop - 8) / OPTION_HEIGHT + LIST_HEIGHT / 2 / 24
    );
    let children = Array.isArray(props.children) ? props.children : [];
    children = children.slice(
      Math.max(0, middleItemIndex - RANGE),
      Math.min(middleItemIndex + RANGE, children.length)
    );
    const filteredChildren = children.filter(
      ({ props: { data } }) => !cacheRef.current[data.family]
    );

    filteredChildren.forEach(({ props: { data: font } }) => {
      appendRegularFontToHead(font);
      cacheRef.current[font.family] = true;
    });
  };

  React.useEffect(() => {
    const handleScroll = debounce((event) => {
      injectStyle(event.target);
    });
    if (menuListRef.current) {
      menuListRef.current.addEventListener("scroll", handleScroll);

      return () => {
        if (menuListRef.current) {
          menuListRef.current.removeEventListener("scroll", handleScroll);
        }
      };
    }
  }, []);

  React.useEffect(() => {
    if (menuListRef.current) {
      injectStyle(menuListRef.current);
    }
  }, [props.children.length]);

  return (
    <components.MenuList
      innerRef={(dom) => {
        menuListRef.current = dom;
        innerRef.current = dom;
      }}
      {...props}
    />
  );
};

const Option = ({ ...props }) => {
  return (
    <components.Option
      {...props}
      innerProps={{
        ...props.innerProps,
        style: { fontFamily: `'${props.data.label}', sans-serif` },
      }}
    />
  );
};

const SingleValue = ({ ...props }) => {
  return (
    <components.SingleValue
      {...props}
      innerProps={{
        ...props.innerProps,
        style: { fontFamily: `'${props.data.label}', sans-serif` },
      }}
    />
  );
};

export const GoogleFontSelect = ({ apiKey, minWidth = 156, ...props }) => {
  const [data, setData] = React.useState([]);
  const currentFont = props.value ? props.value.family : "";
  const htmlLink = React.useRef(document.createElement("link"));

  React.useEffect(() => {
    // storybook is loading iframe
    // contentDocument.head can be null
    function attachHtmlLink(head) {
      if (head) {
        const link1 = document.createElement("link");
        link1.rel = "preconnect";
        link1.href = "https://fonts.googleapis.com";

        const link2 = document.createElement("link");
        link2.rel = "preconnect";
        link2.href = "fonts.gstatic.com";
        link2.setAttribute("crossorigin", true);

        head.appendChild(link1);
        head.appendChild(link2);

        htmlLink.current.rel = "stylesheet";
        head.appendChild(htmlLink.current);
      }
    }
    const callUntil = () => {
      const timeout = setTimeout(() => {
        const iframe = document.getElementById("storybook-preview-iframe");
        if (!iframe.contentDocument.head) {
          callUntil();
        } else {
          attachHtmlLink(iframe.contentDocument.head);
        }
      }, 400);
      return timeout;
    };
    const timeout = callUntil();
    return () => {
      clearTimeout(timeout);
    };
  }, []);

  React.useEffect(() => {
    let mounted = true;
    if (apiKey) {
      fetch(
        `https://www.googleapis.com/webfonts/v1/webfonts?key=${apiKey}&sort=popularity`
      )
        .then((response) => response.json())
        .then((response) => {
          if (response && mounted) {
            setData(response.items || []);
          }
        });
    }
    return () => {
      mounted = false;
    };
  }, [apiKey]);

  React.useEffect(() => {
    if (currentFont) {
      const fontData = data.find(({ family }) => family === currentFont);
      if (fontData) {
        appendRegularFontToHead(fontData);
      }
    }
  }, [currentFont, data.length]);

  React.useEffect(() => {
    if (currentFont) {
      let apiUrl = [];
      apiUrl.push("https://fonts.googleapis.com/css2?family=");
      apiUrl.push(currentFont.replace(/ /g, "+"));
      apiUrl.push(":wght@");
      apiUrl.push([100, 200, 300, 400, 500, 600, 700, 800, 900].join(";"));
      apiUrl.push("&display=swap");

      apiUrl = apiUrl.join("");

      htmlLink.current.href = apiUrl;
    }
  }, [currentFont]);

  const options = data.map(({ family, ...item }) => ({
    family,
    ...item,
    label: family,
    value: family,
  }));
  return (
    <Select
      {...props}
      options={options}
      menuPortalTarget={document.body}
      components={{ Option, SingleValue, MenuList, ...props.components }}
      styles={{
        placeholder: (styles) => ({
          ...styles,
          fontFamily:
            '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
        }),
        valueContainer: (styles) => ({
          ...styles,
          paddingTop: 0,
          paddingBottom: 0,
        }),
        container: (styles) => ({
          ...styles,
          alignSelf: "center",
          minWidth,
          fontSize: 14,
        }),
        control: (styles) => ({
          ...styles,
          minHeight: "auto",
          border: "none",
          backgroundColor: "#efefef",
        }),
        indicatorsContainer: (styles) => ({
          ...styles,
          "& > div": {
            padding: 4,
          },
        }),
        menuList: (styles) => ({
          ...styles,
          maxHeight: LIST_HEIGHT,
        }),
        option: (styles) => ({
          ...styles,
          fontSize: "0.75rem",
          paddingTop: 0,
          paddingBottom: 0,
          height: OPTION_HEIGHT,
          display: "flex",
          alignItems: "center",
        }),
        ...props.styles,
      }}
    />
  );
};
